segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/AppendProcessorTest.java

```diff
@@ -124,6 +136,101 @@ public void testAppend() {
+        assertTrue(processor.isSetupAppendCompleted(setupAppendCommand.getSegment(), setupAppendCommand.getWriterId()));
+    }
+
+    @Test
+    public void testSetupAppendClosesConnectionIfTokenHasExpired() {
+        String streamSegmentName = "scope/stream/0.#epoch.0";
+        UUID clientId = UUID.randomUUID();
+        byte[] data = new byte[] { 1, 2, 3, 4, 6, 7, 8, 9 };
+        StreamSegmentStore store = mock(StreamSegmentStore.class);
+        ServerConnection connection = mock(ServerConnection.class);
+        ConnectionTracker tracker = mock(ConnectionTracker.class);
+        val mockedRecorder = Mockito.mock(SegmentStatsRecorder.class);
+
+        AppendProcessor processor = AppendProcessor.defaultBuilder()
+                .store(store)
+                .connection(connection)
+                .connectionTracker(tracker)
+                .statsRecorder(mockedRecorder)
+                .tokenVerifier(new TokenVerifierImpl("secret"))
+                .build();
+
+        setupGetAttributes(streamSegmentName, clientId, store);
+        val ac = interceptAppend(store, streamSegmentName, updateEventNumber(clientId, data.length), CompletableFuture.completedFuture((long) data.length));
+
+        Date expiryDate = Date.from(Instant.now().minusSeconds(100));
+        JsonWebToken token = new JsonWebToken("subject", "audience", "secret".getBytes(), expiryDate, null);
+
+        SetupAppend setupAppend = new SetupAppend(1, clientId, streamSegmentName, token.toCompactString());
+        processor.setupAppend(setupAppend);
+        verify(connection).close();
+    }
+
+    @Test
+    public void testSetupTokenExpiryTaskClosesConnectionIfTokenHasExpired() throws InterruptedException {
+        // Arrange
+        String streamSegmentName = "scope/stream/0.#epoch.0";
+        UUID clientId = UUID.randomUUID();
+
+        StreamSegmentStore mockStore = mock(StreamSegmentStore.class);
+        ServerConnection mockConnection = mock(ServerConnection.class);
+
+        @Cleanup("shutdown")
+        ScheduledExecutorService executor = new InlineExecutor();
+        AppendProcessor processor = AppendProcessor.defaultBuilder()
+                .store(mockStore)
+                .connection(mockConnection)
+                .tokenExpiryHandlerExecutor(executor)
+                .build();
+
+        // Spy the actual Append Processor, so that we can have some of the methods return stubbed values.
+        AppendProcessor mockProcessor = spy(processor);
+        doReturn(true).when(mockProcessor).isSetupAppendCompleted(streamSegmentName, clientId);
+        // doReturn(Duration.ofMillis(20)).when(mockProcessor);
+
+        JsonWebToken token = new JsonWebToken("subject", "audience", "secret".getBytes(),
+                Date.from(Instant.now().plusMillis(100)), null);
+        SetupAppend setupAppend = new SetupAppend(1, clientId, streamSegmentName, token.toCompactString());
+
+        Thread.sleep(100);
```

Comment:
Please do everything you can to avoid sleeping arbitrary amounts of time in unit tests. This just adds unnecessary delays.

Reply:
Yeah, I agree with you fully. This was supposed to be a stop-gap, until I found a better solution, but I pushed the change by mistake. In any case, the solution turned out to be quite simple, and I have pushed the change in the latest commit.